Core: Speed up playthrough sphere culling

Mysteryem opened the pull request:
## What is this fixing or adding?

To determine if an item in a sphere is required, playthrough calculation would remove the item from its location and then use can_beat_game to sweep through locations until either the game can be beaten (the item is not required) or until the sweep runs out of reachable locations (the item is required).

This patch changes how it is determined if an item is required.

The advancement locations in each higher sphere that have already been culled must be reached for the game to be beatable, so iterate through the advancement locations in each higher sphere after attempting to cull an item. If all locations in a higher sphere and all unreachable locations carried over from a lower sphere are unreachable, then the game is unbeatable and the item is therefore required to beat the game.

For items which are required to beat the game, this aborts faster than exhausting all reachable locations to determine that the game is unbeatable.

For items which are not required to beat the game, the locations which should be reachable from the current sphere are known in advance, so this only has to check reachability of the locations in the closest higher sphere, and any unreachable locations carried over from lower spheres, before moving onto the next closest sphere, rather than repeatedly checking all filled locations which have yet to be reached.

This results in a faster playthrough calculation that also scales better than the previous implementation as the number of advancement items in the multiworld increases.

Playthrough calculation duration averaged over 5 generations of the same seed recorded before/s -> after/s.
57 unique games: 671.2 -> 179.5 (73.3% reduction)
28 unique games: 149.2 -> 47.8 (68.0% reduction)
14 unique games: 7.0 -> 4.2 (39.4% reduction)
7 unique games: 4.6 -> 3.4 (26.8% reduction)

Because the spheres are re-arranged and new spheres are added while iterating, there is no longer a need to calculate the final spheres because the result is already in its final spheres.

The type hint in create_paths has been changed from List[Set[Location]] to Collection[Set[Location]] to allow passing it a deque.

## How was this tested?

Generated with `--seed 1` at d65863ffa2f0 and `spoiler: 2` in host.yaml and averaged over 5 generations.

Playthrough only:
| Yamls | Before/s | After/s | % reduction |
| -- | -- | -- | -- |
| The first 7 non-rom template yamls (down to Bumper Stickers) | 4.6 | 3.4 | 26.8% |
| The first 14 non-rom template yamls (down to DOOM 1993) | 7.0 | 4.2 | 39.4% |
| The first 28 non-rom template yamls (down to Mega Man 2) | 149.2 | 47.8 | 68.0% |
| 1 of each non-rom template (57 players) | 671.2 | 179.5 | 73.3% |

Full generation duration:
| Yamls | Before/s | After/s | Reduction |
| -- | -- | -- | -- |
| The first 7 non-rom template yamls (down to Bumper Stickers) | 7.4 | 6.2 | 16.0% |
| The first 14 non-rom template yamls (down to DOOM 1993) | 11.0 | 8.3 | 24.5% |
| The first 28 non-rom template yamls (down to Mega Man 2) | 168.6 | 66.5 | 60.6% |
| 1 of each non-rom template (57 players) | 762.4 | 268.6 | 64.8% |

Mysteryem committed:
Core: Speed up playthrough sphere culling

To determine if an item in a sphere is required, playthrough calculation
would remove the item from its location and then use can_beat_game to
sweep through locations until either the game can be beaten (the item is
not required) or until the sweep runs out of reachable locations (the
item is required).

This patch changes how it is determined if an item is required.

The advancement locations in each higher sphere that have already been
culled must be reached for the game to be beatable, so iterate through
the advancement locations in each higher sphere after attempting to cull
an item. If one of the locations in a higher sphere is unreachable, then
the game is unbeatable and the item is therefore required to beat the
game.

For items which are required to beat the game, this aborts faster than
exhausting all reachable locations to determine that the game is
unbeatable.

For items which are not required to beat the game, the locations which
should be reachable from the current sphere are known in advance, so
this only has to check reachability of the locations in the closest
higher sphere before moving onto the next closest sphere, rather than
repeatedly checking all filled locations which have yet to be reached.

This results in a faster playthrough calculation that also scales
better than the previous implementation as the number of advancement
items in the multiworld increases.

This new implementation seems to not quite cull as many items on average
as before, but I am unsure if this is an issue with the new
implementation or an issue with the logic of the worlds being generated,
e.g. missing indirect connections causing a region to not be reachable
when it should be, which could cause a required location in a higher
sphere to not be reachable when it should be, which would give a false
positive that an item is required to beat the game.

Playthrough calculation duration averaged over 5 generations of the same
seed recorded before/s -> after/s.
57 unique games: 671.2 -> 133.5 (80.1% reduction)
28 unique games: 149.2 -> 34.9 (76.6% reduction)
14 unique games: 7.0 -> 3.1 (55.9% reduction)
7 unique games: 4.6 -> 2.5 (45.4% reduction)

NewSoupVi commented:
That's so smart god DAYUM

NewSoupVi commented:
Wait, I think I see the problem, let me think about this for one moment

NewSoupVi commented:
Yeah I think this checks out

Idk why there is weird behavior. Would be nice to have sth reproducible there

Mysteryem commented:
> Idk why there is weird behavior. Would be nice to have sth reproducible there

I think I see the reason now, and it's described by the comment in create_playthrough about how the pruning stage could have made certain items dependent on others in the same or later spheres.

I generated a Hat in Time seed (with act randomizer disabled to make the spheres easier to understand) and in the initial sphere candidates, it would collect the item at `Mafia Town - Secret Cave` in sphere 2. The item there was required to beat the game.

Normally, reaching this location requires the `Brewing Hat`, but it is also accessible when you have the `HUMT Access` event that is collected when you can reach the Heating Up Mafia Town act (the Brewing Crate in the way is removed in that act). In the initial sphere candidates, `Mafia Town - Secret Cave` would be reached in sphere 2 because of having `HUMT Access`.

In the original implementation, `HUMT Access` gets culled and `Mafia Town - Secret Cave` moves to sphere 4 of the playthrough where the playthrough would have acquired the `Brewing Hat` in sphere 3.
In this new implementation, `Mafia Town - Secret Cave` must be reachable by sphere 2, so `HUMT Access` does not get culled.

Mysteryem committed:
Restore original behaviour of moving locations to later spheres

Doesn't scale as well as before, but is still a good improvement.

Because the spheres are re-arranged and new spheres are added while
iterating, there is no longer a need to calculate the final spheres
because the result is already in its final spheres.

This should now match the original playthrough implementation.

The type hint in create_paths has been changed from List[Set[Location]]
to Collection[Set[Location]] to allow passing it a deque.

NewSoupVi commented:
Right, that makes sense. So the playthrough won't actually be locally minimal anymore? It could be that one of the items is merely in there to keep a location it its earliest possible sphere?
That would be a sacrifice, but I think it could also have some upsides, because this new method would guarantee the lowest number of spheres (I think), which some might consider more desirable than the playthrough being as minimal as possible. I think some people would prefer this actually.

I definitely think this is worth having, I'm just a bit unsure now as to whether it should be optional, and if yes, which one should be the default. I'd probably lean towards: optional (host.yaml), with the new method being the default, and the old code being called "true minimal". Would love to hear more opinions thought

ScipioWright commented:
I agree with the new method being the default, and the old method kept around for legacy purposes.

Mysteryem commented:
I think I may be able to turn this back into a 'true minimal' with some modifications. It won't be as fast the current 'minimal spheres', but it should still be faster than the original.

The extra things I'm doing in the cull loop are:
1) Instead of immediately aborting when a location is unreachable, add that location to a list to carry over to the next sphere.
2) If no locations in the sphere or any carried over locations were reachable, then abort because I think this means that the game is unbeatable because I think it should not be possible to reach any locations in later spheres.
3) If a carried over location was reachable, write the location and its new sphere to a list. The location will be moved to its new sphere if the game is still beatable.
4) Proceed to the end of the spheres as before and finally check if the game is beaten. If it is, then adjust the spheres of any locations which have changed spheres.

Altering the spheres while culling is important to prevent the next cull attempt from having to carry along the same locations that are no longer reachable in their original spheres, due to a location being culled, but are reachable in a later sphere.

I need to double check if this is correct and clean up the code if it is, but here's how it's looking by comparison:

Playthrough only:
| Yamls | 'True minimal' (main branch)/s | 'Minimal spheres' (current PR)/s | 'True minimal v2' (WIP)/s |
| -- | -- | -- | -- |
| The first 7 non-rom template yamls (down to Bumper Stickers) | 4.6 | 2.5 | 3.4 |
| The first 14 non-rom template yamls (down to DOOM 1993) | 7.0 | 3.1 | 4.3 |
| The first 28 non-rom template yamls (down to Mega Man 2) | 149.2 | 34.9 | 47.4 |
| 1 of each non-rom template (57 players) | 671.2 | 133.5 | 198.0 |

NewSoupVi commented:
A version that's just faster with the same results should obviously be added so I'd definitely wanna see that!!

beauxq commented on BaseClasses.py:
```diff
@@ -1357,34 +1357,80 @@ def create_playthrough(self, create_paths: bool = True) -> None:
         for num, sphere in reversed(tuple(enumerate(collection_spheres))):
-            to_delete: Set[Location] = set()
-            for location in sphere:
+            for location in tuple(sphere):
```
It might be good to include a comment on why this `tuple` is here, so readers don't have to wonder about it for a minute like I did.

Mysteryem commented:
I've restored the original behaviour of allowing locations to move to later spheres and allowing new spheres to be created, and I've updated the PR description.

As expected, it's a little slower and matching the original behaviour is more complicated. I thought it was interesting to note that playthrough can also create entirely new spheres, so I added a debug log when new spheres get added.

NewSoupVi commented:
Nice! Ftr, there does still seem to be some interest in the original version of this, so keep that code around too for a separate PR maybe if you want :3

Mysteryem commented:
It looks like playthrough is another thing that would benefit from knowing which worlds each world logically depends on.

If it is assumed that each world only logically depends on its own items/locations/etc. then immediately after each cull attempt, only the locations belonging to the player the culled item belongs to need to be checked. For each of their locations that became unreachable, the item at that location also becomes unreachable, so the owning player of that item could then also have unreachable locations in a higher sphere, so their locations will also have to be checked from the next sphere onwards.

So, start with a `could_have_unreachable_locations = {culled_item.player}`.
For locations in the sphere (not carried over from previous spheres):
- if `location.player` is not in `could_have_unreachable_locations` then it is assumed the location is reachable
- if `location.player` is in `could_have_unreachable_locations` then `location.can_reach(culled_state)` is checked
  - if it is not reachable then `location.item.player` will be added to `could_have_unreachable_locations` at the end of the current sphere

When it is knowable which worlds each world logically depends on it would be more like:
`could_have_unreachable_locations = players_logically_dependent_on(culled_item.player)`
and
- [...]
  - if it is not reachable then `players_logically_dependent_on(location.item.player)` will be added to `could_have_unreachable_locations` at the end of the current sphere

Mysteryem committed:
Do not remove items from locations when sphere culling

This was unnecessary because locations are removed from their spheres
instead.

This fixes issues that arise from the fact that some worlds have logic
that changes when a specific item is at a specific location.

Mysteryem committed:
Merge remote-tracking branch 'upstream/main' into playthrough_faster_cull

# Conflicts:
#	BaseClasses.py

Mysteryem committed:
Merge remote-tracking branch 'upstream/main' into playthrough_faster_cull

# Conflicts:
#	BaseClasses.py

Mysteryem committed:
Merge remote-tracking branch 'upstream/main' into playthrough_faster_cull

Mysteryem committed:
Re-add verification that the calculated playthrough is valid

Mysteryem committed:
Add verification that the playthrough is able to beat the game

The playthrough provides a path for all players to reach their goals.
If the goals cannot be reached by following the playthrough, then raise
an error.

darkfire006 reviewed (approved):
I use this to generate a sync quickly and effectively every week!

Mysteryem committed:
Merge branch 'main' into playthrough_faster_cull

Mysteryem committed:
Merge branch 'main' into playthrough_faster_cull
